plugin-server/src/worker/plugins/run.ts

```diff
@@ -161,22 +136,85 @@ export async function runPluginTask(
         }
         response = await (payload ? task?.exec(payload) : task?.exec())
     } catch (error) {
-        await processError(server, pluginConfig || null, error)
+        await processError(hub, pluginConfig || null, error)
 
-        server.statsd?.increment(`plugin.task.ERROR`, {
+        hub.statsd?.increment(`plugin.task.ERROR`, {
             taskType: taskType,
             taskName: taskName,
             pluginConfigId: pluginConfigId.toString(),
             teamId: teamIdStr,
         })
     }
-    server.statsd?.timing(`plugin.task`, timer, {
+    hub.statsd?.timing(`plugin.task`, timer, {
         plugin: pluginConfig?.plugin?.name ?? '?',
         teamId: teamIdStr,
     })
     return response
 }
 
-function getPluginsForTeam(server: Hub, teamId: number): PluginConfig[] {
-    return server.pluginConfigsPerTeam.get(teamId) || []
+/** Run function with `RetryError` handling. Returns the number of attempts made. */
+export async function runRetriableFunction(
+    hub: Hub,
+    pluginConfig: PluginConfig,
+    event: ProcessedPluginEvent,
+    tag: string,
+    tryFunc: () => Promise<void>,
+    catchFunc?: (error: Error) => Promise<void>,
+    finallyFunc?: () => Promise<void>
+): Promise<number> {
+    const timer = new Date()
+    let attempt = 0
+    const teamIdString = event.team_id.toString()
+    while (true) {
+        attempt++
+        let nextRetryMs: number
+        try {
+            await tryFunc()
+            break
+        } catch (error) {
+            if (error instanceof RetryError) {
+                error._attempt = attempt
+                error._maxAttempts = ON_CAUSE_MAX_ATTEMPTS
+            }
+            if (error instanceof RetryError && attempt < ON_CAUSE_MAX_ATTEMPTS) {
+                nextRetryMs = ON_CAUSE_RETRY_BASE_MS * ON_CAUSE_RETRY_MULTIPLIER ** (attempt - 1)
+                hub.statsd?.increment(`plugin.${tag}.RETRY`, {
+                    plugin: pluginConfig.plugin?.name ?? '?',
+                    teamId: teamIdString,
+                    attempt: attempt.toString(),
+                })
+            } else {
+                await catchFunc?.(error)
+                await processError(hub, pluginConfig, error, event)
+                hub.statsd?.increment(`plugin.${tag}.ERROR`, {
+                    plugin: pluginConfig.plugin?.name ?? '?',
+                    teamId: teamIdString,
+                    attempt: attempt.toString(),
+                })
+                break
+            }
+        }
+        await delay(nextRetryMs)
+    }
+    await finallyFunc?.()
+    hub.statsd?.timing(`plugin.${tag}`, timer, {
```

Comment:
Is this an existing or a new metric? Slightly worried about overloading influxdb with metrics we won't actually use.

Reply:
This is all existing, only the `.RETRY` one is new, but if we want to keep the rest, I'm strongly in favor of tracking errors _and_ retries.